proto/docs/p4runtime.md

```diff
@@ -515,3 +515,164 @@ messages refer to singleton ports by default.
+will normally expect packet-out metadata (`egress_port`) value of bitwidth as
+defined by the PSA target's `PortIdInHeader_t` type and in the space of the
+target's port numbers. However, the presence of the `@p4runtime_translation`
+annotation will signal to the p4runtime service that it should expect a
+controller defined type and value of `egress_port` metadata. The p4runtime
+service should use the target config to translate this value into PSA
+device-specific value of type `PortIdInHeader_t`.
+
+A similar reverse translation is required in the P4Runtime service for packets
+punted from the target to the controller as shown by the packet-in header
+example below.
+```
+@controller_header("packet_in")
+header PacketIn_t {
+  @p4runtime_translation("port")
+  PortIdInHeader_t ingress_port;
+  @p4runtime_translation("cos")
+  ClassOfServiceInHeader_t cos;
+}
+```
+A packet punted from the target's PSA device will be intercepted by the
+P4runtime service before being sent to the controller. Normally, the P4Runtime
+service will extract the values of `ingress_port` and `cos` from the header and
+insert them in the packet-in metadata fields before sending the packet over the
+stream channel to the controller. However, the presence of the
+`@p4runtime_translation` annotations will be a signal to the P4Runtime service
+to replace these values with equivalent controller-defined numbers from the
+target config.
+
+### Translation of Match Fields
+Port and class of service type match fields can be part of a P4 table's match
+key as shown in the example below:
+```
+table t {
+  key = {
+    istd.ingress_port: exact @p4runtime_translation("port");
+    istd.class_of_service: ternary @p4runtime_translation("cos");
+  }
+  actions = {
+    drop;
+  }
+}
+```
+Table `t` has an `exact` match on ingress port and `ternary` match on class of
+service metadata. A P4Runtime write request from the controller will have
+the values of the match fields set to controller-specific values. The
+`@p4runtime_translation` annotation on the match fields is a signal to the
+P4Runtime service to intercept these write requests and use the configuration
+data to translate the values to respective device-specific values. Similarly,
+the P4Runtime service should translate the device-specific values to
+controller-specific values when returning the read response for table `t`.
+
+Note that it may be infeasible to translate the value-mask pair for `ternary`
+matches and the value-prefix-length pair for `lpm` matches. Therefore, the
+P4Runtime service may require that the match be effectively either exact (all-1
+mask for ternary and prefix-length of the field's controller-bitwidth for lpm)
+or don't care (0 mask for ternary and 0 prefix length for lpm).
+
+### Translation of Action Parameters
+Port and class of service type parameters can be part of a P4 action definition
+as shown in the example below:
+```
+action a(@p4runtime_translation("port") PortId_t p,
```

Comment:
Looking at this example, I am wondering why the controller-chosen bitwidth is not part of the annotation (e.g. `@p4runtime_translation("port", 32)`. This information would be needed in the P4Info to sanity-check client messages.

In a way this is related to @teverman's question above. Is the controller always sending 64-bit bitstrings? If yes, that doesn't match the example you gave earlier in the doc.

Reply:
@antoninbas Are you expecting different controllers to choose different bitwidths for these quantities in the Protobuf messages they send/receive?  If so, for what purpose would they choose different bitwidths?  64 bits should be wide enough for all such occurrences of values of these types, so my first guess is that the motivation would be to economize on bits sent over the wire between client and server?

In any case, I don't see why the bitwidth of these fields between client and server should affect the P4 source code?

Reply:
Yes, this was a recent decision. I will update the doc. The controller's port bitwidth is known through the switch config (outside of P4Runtime). It really is not needed in this annotation.

Reply:
In P4Info, you really just need a bool to indicate that the field requires translation. I think even that is just a convenience since the annotation will show up anyway. But such a flag will make the switch code more efficient since it won't have to parse the annotations for every field. I can leave that for a different PR if needed.

Reply:
Sorry I just came back from traveling.

I don't believe it is a good idea to introduce that kind of strong asymmetry between regular "fields" and translated "fields". Unless I am mistaken, with your proposal, either there is no more bitwidth information for the field in the p4info or the bitwidth information is meaningless / misleading because it doesn't match what we expect from the controller.

Even if we put aside sanity-checking on the server-side, I believe one should be able to auto-generate a strongly-typed control-plane API based on p4info only, but I don't see how I can do that for translated fields if I'm missing the bitwidth information?

Since this information should be available at compile-time and is not expected to be dynamic, I am not sure why it cannot be included in the annotation & p4info. What's the rationale behind your statement "It really is not needed in this annotation.". Because I can think of situations where it is actually needed or useful to have.

Reply:
Unless as @jafingerhut wrote, we agree that all translated fields should have bitwidth 64 bit in p4info and be represented as 8-byte bytestrings in p4runtime, which works for me (but I don't see that anywhere in your proposal, even after your most recent changes).

Reply:
+1 to Antonin's comments.

I was trying to think of how to express my concern about the indeterminate bitwidth information.

I am also OK with 64 bits and 8-byte strings for the translated fields, but it needs to be documented cleanly.  I really don't like the idea of bitwidths being a config parameter.

Reply:
I don't think so. The switch configuration is a configuration of the P4Runtime service as well. We should not be having two different "SDN port bitwidths" defined: one in the config and one in the P4 source. In fact, I am also not very fond of the uint64 types we have for ports/cos in the PRE API. But we can live with them because there is no global field in P4Info that tells us the PSA port/cos P4 bitwidth.

The implementation of P4Runtime service is quite clear:
if annotation: use config bitwidth and translate
else: no translate

Are we worried that the switch may not have a configuration outside P4? If so, a program with such annotations should be rejected by the switch (or the annotations may be ignored).

Reply:
Couple of other points:
1. The bitwidth in p4info is only to tell the P4 type (needed if the SDK talks in terms of HW ports only) 
2. The control plane API (I believe you are referring to PD-PI) will know the controller port API also through a similar config that the switch does. And these configs are outside the scope of the P4 program.

My main concern here is that there will be N places for defining SDN types instead on 1. Also, a one-size fits all approach of 64 bit types contradicts a potentially 32-bit type in the config.

Reply:
@teverman @antoninbas Can we try to wrap up this PR for Version 1.0? Thanks.

Reply:
Getting this in for v1.0 is definitely the plan, but I don't think we have resolved this discussion yet. Maybe we can discuss it tomorrow at the WG meeting after the type serialization discussion if there is enough time? Otherwise we can always schedule an extra hangout meeting later in the week.